homeassistant/components/azure_data_explorer/__init__.py

```diff
@@ -0,0 +1,225 @@
+    await adx.async_start()
+    return True
+
+
+async def async_update_listener(hass: HomeAssistant, entry: ConfigEntry) -> None:
+    """Update listener for options."""
+    hass.data[DOMAIN][DATA_HUB].update_options(entry.options)
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    adx = hass.data[DOMAIN].pop(DATA_HUB)
+    await adx.async_stop()
+    return True
+
+
+class AzureDataExplorer:
+    """A event handler class for Azure Data Explorer."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        entry: ConfigEntry,
+    ) -> None:
+        """Initialize the listener."""
+
+        self.hass = hass
+        self._entry = entry
+        self._entities_filter = hass.data[DOMAIN][DATA_FILTER]
+
+        self._client = AzureDataExplorerClient(
+            CONF_ADX_CLUSTER_INGEST_URI=self._entry.data["clusteringesturi"],
+            CONF_ADX_DATABASE_NAME=self._entry.data["database"],
+            CONF_ADX_TABLE_NAME=self._entry.data["table"],
+            CONF_APP_REG_ID=self._entry.data["client_id"],
+            CONF_APP_REG_SECRET=self._entry.data["client_secret"],
+            CONF_AUTHORITY_ID=self._entry.data["authority_id"],
+            CONF_USE_FREE=self._entry.data["use_free_cluster"],
+        )
+
+        self._send_interval = self._entry.options[CONF_SEND_INTERVAL]
+        self._max_delay = DEFAULT_MAX_DELAY
+
+        self._shutdown = False
+        self._queue: asyncio.PriorityQueue[
+            tuple[int, tuple[datetime, State | None]]
+        ] = asyncio.PriorityQueue()
+        self._listener_remover: Callable[[], None] | None = None
+        self._next_send_remover: Callable[[], None] | None = None
+
+    async def async_start(self) -> None:
+        """Start the component.
+
+        This register the listener and
+        schedules the first send.
+        """
+
+        self._listener_remover = self.hass.bus.async_listen(
+            MATCH_ALL, self.async_listen
+        )
+        self._schedule_next_send()
+
+    async def async_stop(self) -> None:
+        """Shut down the ADX by queueing None, calling send, join queue."""
+        if self._next_send_remover:
+            self._next_send_remover()
+        if self._listener_remover:
+            self._listener_remover()
+        await self._queue.put((3, (utcnow(), None)))
+        await self.async_send(None)
+        await self._queue.join()
+
+    def update_options(self, new_options: dict[str, Any]) -> None:
+        """Update options."""
+        self._send_interval = new_options[CONF_SEND_INTERVAL]
+
+    async def test_connection(self) -> None:
+        """Test the connection to the Azure Data Explorer service."""
+        await self.hass.async_add_executor_job(self._client.test_connection)
+
+    def _schedule_next_send(self) -> None:
+        """Schedule the next send."""
+        if not self._shutdown:
+            self._next_send_remover = async_call_later(
```

Comment:
```suggestion
            if self._next_send_remover:
                 self._next_send_remover()
            self._next_send_remover = async_call_later(
```

If `async_send` is called manually, it is possible that we still have scheduled another call.